crates/router/src/lib.rs

```diff
@@ -17,72 +18,96 @@ impl Guest for Component {
+        let base = manifest
+            .application
+            .trigger_global_configs
             .get("http")
-            .unwrap()
-            .iter()
-            .map(|trigger| {
-                let spin_manifest::schema::v2::ComponentSpec::Reference(comp) =
-                    trigger.component.as_ref().unwrap()
-                else {
-                    todo!()
-                };
-                (
-                    trigger
-                        .config
-                        .get("route")
-                        .and_then(|route| route.as_str())
-                        .unwrap(),
-                    comp.to_string(),
-                )
-            })
-            .collect::<Vec<_>>();
-        let path_with_query = request
-            .path_with_query()
-            .unwrap_or_else(|| String::from("/"));
-        let path = path_with_query
-            .split_once('?')
-            .map(|(path, _)| path)
-            .unwrap_or(&path_with_query);
-        let routing = routes
-            .iter()
-            .find(|(route, _)| path.starts_with(route) || *route == "/...");
-        if let Some((route, component_id)) = routing {
-            let request = match apply_request_transformations(request, (*route).to_owned()) {
-                Ok(request) => request,
-                Err(e) => {
-                    ResponseOutparam::set(
-                        response_out,
-                        Err(ErrorCode::InternalError(Some(e.to_string()))),
-                    );
-                    return;
-                }
+            .and_then(|c| c.get("base").and_then(|v| v.as_str()))
+            .unwrap_or("/");
+        let route_match = match find_matching_route(&manifest, &request, base) {
+            Err(e) => {
+                set_error_response(response_out, e.to_string());
+                return;
+            }
+            Ok((_path, Some(routing))) => routing,
+            Ok((path, None)) => {
+                set_error_response(
+                    response_out,
+                    format!("no route found in spin.toml manifest for request path '{path}'"),
+                );
+                return;
+            }
+        };
+        let request = match apply_request_transformations(request, base, &route_match) {
+            Ok(request) => request,
+            Err(e) => {
+                set_error_response(response_out, e.to_string());
+                return;
+            }
+        };
+        bindings::set_component_id(route_match.component_id());
+        downstream(request, response_out)
+    }
+}
+
+fn set_error_response(response_out: ResponseOutparam, message: impl Into<String>) {
+    ResponseOutparam::set(
+        response_out,
+        Err(ErrorCode::InternalError(Some(message.into()))),
+    );
+}
+
+fn find_matching_route(
+    manifest: &spin_manifest::schema::v2::AppManifest,
+    request: &IncomingRequest,
+    base: &str,
+) -> anyhow::Result<(String, Option<RouteMatch>)> {
```

Comment:
Nit: The return type here is a bit hard to grok without reading the code. In particular, I find it off that the "happy path" of the `RouteMatch` being found is the second item of the tuple.
Perhaps this is a good example of where a bespoke enum would be useful for the return type since it's not clear what the String represents and when the Option could be none without reading the code fully.

Perhaps:

```rust
enum RoutingResult {
  RouteFound(RouteMatch),
  RouteNotFound { path: String }
}
```